test/fx/test_cse_pass.py

```diff
@@ -0,0 +1,240 @@
+        true_result = fx_g(t)
+        our_result = new_g(t)
+        if true_result is None:  # both return None
+            assert our_result is None, f"true result is None, CSE result is {our_result}"
+        else:  # results returned are the same
+            assert torch.all(true_result == our_result), (
+                f"results are different {true_result}, {our_result}")  # check results are the same
+
+
+class NoChangeTestCase(TestCase):
+
+    def test_nochange(self):
+        def f(x):
+            a = x + 1
+            b = x + a
+            a = x
+            d = x + a
+            return b + d
+        t = torch.randn(2, 2)
+        check(f, t, 0)
+
+    def test_empty(self):
+        def f(x):
+            pass
+        t = torch.randn(2, 2)
+        check(f, t, 0)
+
+
+class ReduceTestCase(TestCase):
+
+    def test_immutable_list_type(self):
+        def f(x):
+            a = x.sum(dim=1)
+            b = x.sum(dim=1)
+            c = x.sum()
+            d = x.sum()
+            return a + b + c + d
+        t = torch.randn(2, 2)
+        check(f, t, 2)
+
+    def test_immutable_list_multiple_entries(self):
+        def f(x):
+            a = x.sum(dim=[0, 1])
+            b = x.sum(dim=[0, 1])
+            c = x.sum(dim=1)
+            d = x.sum(dim=1)
+            return a + b + c + d
+        t = torch.randn(2, 2)
+        check(f, t, 2)
+
+    def test_simple(self):
+        def f(x):
+            a = x.cos()
+            b = x.cos()
+            c = a + a
+            d = b + b
+            return c + d
+        t = torch.randn(2, 2)
+        check(f, t, 2)
+
+    def test_simple_2(self):
+        def f(x):
+            a = x.cos().sin()
+            b = x.cos().sin()
+            c = a + a
+            d = b + b
+            return c + d
+        t = torch.randn(1)
+        check(f, t, 3)
+
+    def test_two_args_default(self):
+        def f(x):
+            a = x.sum(dim=1)
+            b = x.sum(dim=1, keepdim=False)
+            c = x.sum(dim=1, keepdim=False)
+            d = x.sum(dim=1)
+            return a + b + c + d
+        t = torch.randn(2, 2)
+        check(f, t, 3)
+
+    def test_two_args(self):
+        def f(x):
+            a = x.sum(dim=1)
+            b = x.sum(dim=1, keepdim=True)
+            c = x.sum(dim=1, keepdim=True)
+            d = x.sum(dim=1)
+            return a + b + c + d
+        t = torch.randn(2, 2)
+        check(f, t, 2)
+
+    def test_simple_multiple_same_ops(self):
+        def f(x):
+            a = x.sum()
+            b = x.sum()
+            c = x.sum()
+            d = x.sum()
+            return a + b + c + d
+        t = torch.randn(2, 2)
+        check(f, t, 3)
+
+    def test_nested_immutable_list_type(self):
+        def f(x):
+            a = torch.cat((x, x))
+            b = torch.cat((x, x))
+            return a + b
+        t = torch.randn(2, 2)
+        check(f, t, 1)
+
+    def test_kwarg(self):
+        def f(x):
+            a = torch.ones_like(x)
```

Comment:
One interesting pattern to consider if tensor meta is present. @Chillee 

```
def f(x):
   y = x + 1
   a = torch.one_like(x)
   b = torch.one_like(y).  # can we eliminate b in the case, since y and x has the same meta 
```

Reply:
hrmm... it's possible in principle, but I somewhat doubt it matters.